### ISSUE TITLE
Task Takes Too Long

### ISSUE THREAD

**Issue by ldimitroff** [ben-manes/gradle-versions-plugin#505]
I've been using this tool a long time now. And I've noticed that it now takes really long time to check dependencies. 
I haven't run this tool for a while now, but today took 4 hours to complete the task 

```
Generated report file build/dependencyUpdates/report.txt

Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/6.1.1/userguide/command_line_interface.html#sec:command_line_warnings

BUILD SUCCESSFUL in 4h 11m 4s
```
Am I missing a flag or configuration?
Thanks

**Comment by ben-manes** [comment]
That is unusual. We resolve through Gradle which has rich dependency management, so rather than poorly reimplement it we delegate and report. The most common issue is that a repository is very slow and, since Gradle still resolves sequentially, this can cause it to take too long. The `--info` should help identify the problem and you can use `--refresh-dependencies` to disable the local dependency cache. When identified, you can use [repository content filtering](https://docs.gradle.org/current/userguide/declaring_repositories.html#sec:repository-content-filtering) to narrow the repository's usage, e.g. only JS dependencies should resolve against npm.

**Comment by ldimitroff** [comment]
I just ran the task with the `--info` option and this is the output: 

```Selection of androidx.appcompat:appcompat:1.1.0-rc01 rejected by component selection rule: Release candidate
Resource missing. [HTTP HEAD: https://jcenter.bintray.com/androidx/appcompat/appcompat/1.1.0-rc01/appcompat-1.1.0-rc01.pom]
Resource missing. [HTTP HEAD: https://jitpack.io/androidx/appcompat/appcompat/1.1.0-rc01/appcompat-1.1.0-rc01.pom]
Resource missing. [HTTP HEAD: https://repo.clojars.org/androidx/appcompat/appcompat/1.1.0-rc01/appcompat-1.1.0-rc01.pom]
Failed to get resource: HEAD. [HTTP HTTP/1.1 403 Forbidden: https://dl.bintray.com/guardian/android/androidx/appcompat/appcompat/1.1.0-rc01/appcompat-1.1.0-rc01.pom)]
Resource missing. [HTTP HEAD: https://repo.maven.apache.org/maven2/androidx/appcompat/appcompat/1.1.0-rc01/appcompat-1.1.0-rc01.pom]
Resource missing. [HTTP HEAD: https://www.jitpack.io/androidx/appcompat/appcompat/1.1.0-rc01/appcompat-1.1.0-rc01.pom]
Resource missing. [HTTP HEAD: http://repo.brightcove.com/releases/androidx/appcompat/appcompat/1.1.0-rc01/appcompat-1.1.0-rc01.pom]
```

Seems that it's trying every repository declared to get each dependency... 
The sync or updating dependencies does not takes 4 hours  so that's why I'm seing a difference here.

**Comment by ben-manes** [comment]
Repositories are defined at the project level which configurations are declared on, and we resolve all configurations which hold the dependencies and resolution rules. Since gradle doesn't resolve metadata in parallel it can be quite slow. For project dependencies on fixed versions it downloads the artifacts in parallel so the build feels fast. They simply have kept backlogging the metadata optimization because it only impacts us.

I believe bintray was going to introduce a delay to wean users off. Without timing it is not obvious which is the cause. If you use repository filtering then you can avoid queries that make no sense which should speed it up again.

**Comment by henrik242** [comment]
That task length isn't unusual:
```BUILD SUCCESSFUL in 2h 1m 19s```

Would it be possible to log some statistics of the most time consuming tasks/repositories/artifacts?

**Comment by ldimitroff** [comment]
@ben-manes @henrik242 I've fixed locally the issue by removing a snapshot repository before running the dependency updates task.
Noticed that the task was trying to get the information about every verion listed in that repository. It was going decremental instead of checkging the latest version :man_shrugging: 

```
tasks.dependencyUpdates.doFirst {
    project.repositories.removeAll {
        it instanceof MavenArtifactRepository && it.url.toString() == '<MY.SNAPSHOT.REPO>'
    }
}
```

**Comment by ben-manes** [comment]
> Would it be possible to log some statistics of the most time consuming tasks/repositories/artifacts?

@henrik242 Unfortunately we can't, because the resolution is handled by Gradle itself. We don't reimplement their dependency management, but plug into it via a cloned configuration and rewritten dependency versions. Since the APIs don't expose that data, we couldn't provide hints beyond looking at the `--info` or `--debug` logs to see Gradle's messaging.

> Noticed that the task was trying to get the information about every version listed in that repository. It was going decremental instead of checking the latest version

@ldimitroff This is how a resolution strategy works to resolve, where it pulls all of the versions (typically from `maven-metadata.xml`), orders in descending order, and then filters by the component selection rule until a version passes (or fails as unresolvable). If Gradle is making a fetch per version then that would be slow. It may be that if you declare the repository as [snapshot only](https://docs.gradle.org/current/userguide/declaring_repositories.html#maven_repository_filtering), gradle might skip over it. However, since we use a dynamic query (`+`) it might try regardless so your trick is excellent.

**Comment by henrik242** [comment]
I have now removed an unused snapshot repo and sprinkled other repository definitions with stuff like ...

```
            mavenContent {
                releasesOnly()
            }
            content {
                excludeGroupByRegex("com\\.mycompany.*")
            }
```

... and voila! the build times are down to mere minutes :)

**Comment by oheyadam** [comment]
Hey, I'm also experiencing really long times, but not sure if it's for the same reason. I get a lot of the following. Not really sure what `Copy2` is, or whether it's supposed to resolve the dependencies of different build flavors (MockDebug) being one of them.

```
:dependencyUpdates > Resolve dependencies of :app:kotlinCompilerPluginClasspathDevDebugCopy2 > maven-metadata.xml
:dependencyUpdates > Resolve dependencies of :feature:city:apiCopy2 > maven-metadata.xml
:dependencyUpdates > Resolve dependencies of :feature:city:apiCopy2 > maven-metadata.xml
:dependencyUpdates > Resolve dependencies of :feature:city:kotlinCompilerPluginClasspathMockDebugUnitTestCopy2 > maven-metadata.xml
```

**Comment by ben-manes** [comment]
We iterate over all configurations, call the copy() method, muck with dependency versions, and resolve it. Gradle increments a counter on name conflicts. You might look at the info log to see what it is doing

**Comment by oheyadam** [comment]
Getting a lot of this, but not entirely sure how that affects the task runtime so much. It also takes hours on my end

```
Skipping configuration :feature:city:mockDebugAndroidTestCompileClasspath
org.gradle.api.artifacts.UnknownConfigurationException: Configuration with name 'mockDebugAndroidTestCompileClasspathCopy' not found.
```

**Comment by ben-manes** [comment]
That’s weird. Android and Kotlin are not good citizens on Gradle, so it’s hard to guess what is going on from that. After we get project.congratulations we call.
https://github.com/ben-manes/gradle-versions-plugin/blob/8f1f32b3dd0190753bf5142cfe62dbb37c85f156/src/main/groovy/com/github/benmanes/gradle/versions/updates/Resolver.groovy#L128

There shouldn’t be anything magical going on to explain that error.

**Comment by jhansche** [comment]
I have a library project that's ~150 modules, and while it's in development we use SNAPSHOT versions. When a host application is using a snapshot version of those modules, the `dependencyUpdates` task takes 15-20 hours to run.

Running with `--info`, we see that it has to download every .pom and .module file (plus maven-metadata.xml), for the last 25 versions (including both snapshots and releases).  At 150 modules, that's 3750 separate versions that are being evaluated, and 7500 files (pom and module; even more if you include maven-metadata), and it has to do this sequentially.

Is there any way to exclude a groupId from evaluation, for example? I know we can use component selection rules to determine whether a component is acceptable or not - but I want to just completely ignore all checks for our internal groupId. Is that possible?

I could switch the repository that's used when running dependencyUpdates, so that it only considers the releases repository and ignores the snapshot repository.  But when the app is currently using a snapshot, that would prevent it from accessing that snapshot to begin with.

Looking for any other ideas for reducing this runtime. 20 hours is too long.

**Comment by ben-manes** [comment]
You can use repository content filtering to reduce where Gradle queries for dependencies. That usually has a big impact.

Generally one applies the plugin to the root project to aggregate the report. If you apply it to subprojects then I think those could be run independently. A long-term change the Gradle team wants is for us to use that and aggregate those results, which they went through a lot of pain migrating everyone using jacoco to. That would be good for us except this is an old plugin which is being maintained, so we rely on contributions for most activity.

While this plugin uses the Gradle team's preferred approach, you might consider others that re-implement the dependency resolution. That will be more brittle, but benefits by not being tied to their performance bottlenecks. The Gradle team is working with Github on improving Dependabot over the coming moths. There is also refreshVersions which migrated off this plugin for the purpose of faster resolution for the common case, which might match your needs.

The right fix is for Gradle to finally parallelize this, which Maven has done for over a decade. That would be a nice platform change that would benefit the whole community, but anything related to dependency updates has been low priority to them. Since we've been asking for a decade for this fix, it's not something to plan on them getting to and (at least historically when I last tried) were not accepting of external PRs.

**Comment by jhansche** [comment]
We do run the plugin on the root project, but in the context of an application that's consuming these snapshots, it would still have to do those lookups for every snapshot version when we don't really want the plugin to consider that library specifically, since the library is our own and we maintain that version separately. What we mainly want the plugin to consider is everything else: 3rd party dependencies.

The repository that it's hitting is our own internal repository, and we use artifactory virtual repo for that, which proxies everything else as a remote repo (including maven central, google's repo, jitpack, etc). So it's not an issue of hitting dependencies that just aren't there -- they're there, it's just slow because it's looking at the past 25 artifacts (for some reason?) for every single one, only to report all of them under the "Failed to determine the latest version for the following dependencies" section.

Using exclusiveContent to filter which groups go to which repository would only help if it hits a bunch of repositories that will never have those dependencies. But that's not what's happening here.

What's more confusing is, let's say the app is referencing version `5.3.0-SNAPSHOT`; for some reason running the task with `--info` shows that it is going back _several_ versions (i.e., 5.2.0-SNAPSHOT, 5.1.0-SNAPSHOT, 4.23-SNAPSHOT, etc). If the goal is to find updates, why does it have to look back at previous versions? Wouldn't it be more important to restrict the copied configuration to "<current version> _or newer_"? It should only be looking at 5.3.0 and newer, so it doesn't make sense why it has to download every other version that came before the current version.

I'll see if I can update the CI job that runs this report to use a composite build (--include-build) for the internal library project. That way at least it should translate those snapshot versions into composite project dependencies, and hopefully won't spend so much time checking thousands of versions that are irrelevant to the end goal.

**Comment by ben-manes** [comment]
Gradle will inspect older versions if the resolution strategy rejects a version. The `rejectVersionIf` is just a DSL for the common case. Whenever rejecting it is good to provide a cause to assist in the logging. Do you have a rule that is overly broad?

**Comment by jhansche** [comment]
A couple things: our rejectVersionIf implementation is probably old, but what we use is:
```
dependencyUpdates {
    checkConstraints = true
    rejectVersionIf { selection ->
        ['alpha', 'beta', 'rc', 'cr', 'm'].any { qualifier ->
            selection.candidate.version ==~ /(?i).*[.-]${qualifier}[.\d-]*/
        }
    }
}
```
And we also run with `-Drevision=release`. That's because for 3rd party dependencies, we do have a general rule that we want to use stable versions, but obviously for internal dependencies, we do need to use snapshots during development. We have a separate check that rejects using `+` or `SNAPSHOT` for any release artifacts.

But what I'm proposing is not just for `rejectVersionIf` - that DSL is used for each potential candidate that Gradle comes across, and we can choose to accept or reject each one. My proposal would be to completely strip all dependency declarations from the copied configuration, before we even ask Gradle to resolve it.

```
dependencies {
  implementation("com.mycompany.group:lib1:5.3.0")
  implementation("com.mycompany.group:lib2:5.3.0-SNAPSHOT")
  implementation("androidx.appcompat:1.0.0")
}
```
What I'm suggesting is a way to cause the _copied_ `implementation` configuration to have no reference to the filtered modules at all, so it would contain only "androidx.appcompat" using the above example. If we know that we are managing all updates for `com.mycompany.group` internally, then there's no reason for the plugin (or gradle) to attempt to resolve those during this task, because we know exactly where we stand with those.

I guess I could take a similar approach like what's mentioned here <https://github.com/ben-manes/gradle-versions-plugin/issues/505#issuecomment-901927030> and exclude all of those dependencies from all configurations during `dependencyUpdates.doFirst`. 🤔   I'll give that a shot tomorrow.

**Comment by ben-manes** [comment]
> We have a separate check that rejects using + or SNAPSHOT for any release artifacts.

This plugin uses `+` to dynamically resolve the latest version. Do you think that check is interfering with the task? As we resolve your configurations, any custom resolution rules would be applied. The simple approach is to disable those restrictions if this task is executed using either of the two options below. This would likely explain your slow execution times as it rejects all versions.

Option 1:
```groovy
gradle.taskGraph.whenReady { taskGraph ->
  if (!taskGraph.hasTask(dependencyUpdates)) { ... }
}
```

Option 2:
```groovy
if ('dependencyUpdate' !in gradle.startParameter.taskNames) { ... }
```

> My proposal would be to completely strip all dependency declarations from the copied configuration, before we even ask Gradle to resolve it.

A dependency filter would be fine and could to be added if you'd like to work towards that. The [Resolver](https://github.com/ben-manes/gradle-versions-plugin/blob/master/src/main/groovy/com/github/benmanes/gradle/versions/updates/Resolver.groovy) class handles all of the resolutions. We resolve the configuration for the current dependendencies, copy it, rewrite to dynamic versions, resolve, and then report on the comparison. In the report we include up-to-date mostly because suppressing a dependency could make it harder to debug why an update was not shown. If you wanted some filter to not exclude from resolution then I'd ask for it to be reported on for completeness and debuggability. However, I have not been inclined to add it yet because each time the root cause was something else that made sense to fix instead. So instead of adding features it's been better to figure out what is going wrong and determine the right solution afterwards.

> And we also run with -Drevision=release.

The `revision` flag is an artifact of Gradle 1.0 which was based on Apache Ivy. The only options we had was Ivy's revision categories for dynamic lookups, e.g. `revision.latest`. Gradle eventually wrote their own dependency management backend, so for backwards compatibility we [translate](https://github.com/ben-manes/gradle-versions-plugin/blob/b53e44f1459e7950706e77007b916ee522684baf/src/main/groovy/com/github/benmanes/gradle/versions/updates/Resolver.groovy#L222-L241) `revision` to a resolution strategy. However that flag is not very useful otherwise because Apache Ivy dependencies never became popular, and Maven's dependency metadata does not distinguish between milestone and release versions. Therefore you only need to specify it if you want SNAPSHOTs to be included, else it is an ignorable feature.

**Comment by jhansche** [comment]
> Do you think that check is interfering with the task?

No, sorry, I didn't explain that clearly. The check is not Gradle related. It's just a :dependencies report and a simple `grep`, as part of our release process, that will fail to release if there's a + or snapshot in the report. It should not be the cause of this issue (I can certainly double check that tomorrow)

> So instead of adding features it's been better to figure out what is going wrong and determine the right solution afterwards.

I agree with you. It would be better to find the root cause of the slowness and resolve that.

In that regard, one thing I noticed is that there seems to be about a 1-second pause between resolving each pom/module pair, even if that file is already cached locally. I don't have the console output handy right now, but I can follow up tomorrow with more details. From memory, it's something like: "downloading http://artifactory/group/lib/lib.pom", "downloading ...lib.module", "cached checksum matches", ~1 second pause. The --info output doesn't show anything further to explain where the time is being spent here, but I would expect it to be able to run through these artifacts fairly quickly.

Your explanation that gradle should stop looking at other versions when it finds one that matches, could be the problem in seeing, so I'll take a look at that to make sure that my resolution rules aren't inadvertently causing it to perpetually reject all versions of our libraries.

As for using `+` in the copied configuration to identify available updates: do you think it would make more sense to use a version range instead of +? That is, if the current dependency is using version `2.1.0`, what's the reason for setting that to `+`, vs maven style `[2.1.0,)`?

**Comment by ben-manes** [comment]
> As for using + in the copied configuration to identify available updates: do you think it would make more sense to use a version range instead of +? That is, if the current dependency is using version 2.1.0, what's the reason for setting that to +, vs maven style [2.1.0,)?

I believe the nicer approach here would be #69 to show the breadcrumbs (like we do for gradle releases). Then we'd need to resolve `2.1.+`, `2.+`, and `+` to show the upgrade options. If we only did `2.+` then users would not know of a major version, which is worse. They can of course decide against the major version by using a resolution rule to suppress those options and make that decision explicit in their build. The problem with breadcrumbs is that it adds more dependencies to resolve which impacts your slowness issue.

> I would expect it to be able to run through these artifacts fairly quickly.

I recall Artifactory Cloud being unbelievably slow, but that was many years ago since I used it. I think we had to set up our own instance, so if you use that it may be related. In my [original ticket](https://issues.gradle.org/browse/GRADLE-2582) on this issue I measured it adding 5 minutes to a build. Just FYI that it really could be something that dumb.

In #98 Square had a similar setup to yours. You could rewrite the repositories to redirect to Central for the updates checks but still fetch from your internal repository for the artifact during a build.

**Comment by jhansche** [comment]
It's a self-hosted artifactory server in the same subnet (AWS vpc) as the build servers, not jfrog.io cloud hosted artifactory.

**Comment by jhansche** [comment]
I've confirmed that the problem of searching older versions of the snapshot is due to the `-Drevision=release` selection:
```
Selection of com.mycompany:platform:5.4.0-SNAPSHOT rejected by component selection rule: Component status integration rejected by revision release
```
And it looks like this revision rejection happens before our `rejectVersionIf` action is able to inspect it. So I will update that property to `integration` so that my rules will take effect, and check for 3rd-party snapshots separately. This appears to work, and brings the runtime down to 20 minutes total (which is still slow, but better than 20 hours!)

--

As for the slow delay between downloading artifacts (which are mostly already cached, btw):
```
Downloading https://artifactory.mycompany.com/artifactory/repo/com/mycompany/lib/3.0.0-SNAPSHOT/maven-metadata.xml to /Users/jhansche/.gradle/.tmp/gradle_download9927126184123920168bin
Cached resource https://artifactory.mycompany.com/artifactory/repo/com/mycompany/lib/3.0.0-SNAPSHOT/lib-3.0.0-20211217.225254-9.pom is up-to-date (lastModified: Fri Dec 17 17:56:21 EST 2021).
Cached resource https://artifactory.mycompany.com/artifactory/repo/com/mycompany/lib/3.0.0-SNAPSHOT/lib-3.0.0-20211217.225254-9.module is up-to-date (lastModified: Fri Dec 17 17:56:21 EST 2021).
```
This gets logged for every version that is newer than the current version (including snapshots), with about a 1 second delay between each. The delay seems to be consistent, whether the pom/module files were listed as "Cached..up-to-date", or had to be downloaded, so I'm not sure where this delay is coming from.

When I run it with a build scan, I can see in the performance tab that 21 minutes was in "time spent on downloads", and it seems to be downloading in the range of 400 B/s - 14Kb/s, which is atrocious. I'm going to work with our SRE team to investigate that. I sent you the build scan URL privately.

**Comment by ben-manes** [comment]
That's great. I would try the approach in #98 to add the Maven Central repository only for running the dependencyUpdates task, which might avoid waiting on your SREs if it solves the performance problem. You could add the repository as a [metadata source](https://docs.gradle.org/current/userguide/declaring_repositories.html#sec:supported_metadata_sources) rather than an artifact one.

**Comment by ben-manes** [comment]
Usually the difference between runs is whether Gradle is reading from its cache, which `--refresh-dependencies` will bypass. Gradle's lack of parallel resolution is frustrating and they've ignored it for 10 years, but it is also the right fix rather than an ad hoc workaround in the plugin. Please keep pinging them about it or for them to embed this feature natively. I am sympathetic but they are the ones to poke.

**Comment by MartinX3** [comment]
Just to link the task to this issue
https://github.com/gradle/gradle/issues/15071

**Comment by bh-tt** [comment]
Could part of the problem be that the task tries to resolve '+' version for all dependencies in the entire graph, as opposed to only direct dependencies?

I tried this with a Quarkus project (which pulls a very large number of dependencies in, despite only having around 10 or so direct dependencies) and the resolve took over 3 minutes. Similar projects with fewer total dependencies but a similar number of direct dependencies only took about 40 seconds.

https://github.com/ben-manes/gradle-versions-plugin/blob/8c1ee4ba3f78f2791331670ae1b388c7eaf9abc2/gradle-versions-plugin/src/main/kotlin/com/github/benmanes/gradle/versions/updates/Resolver.kt#L94 does seem to indicate it attempts to resolve all dependencies instead of only direct dependencies.

I'm not sure if this could break some builds so I think a configuration option to only resolve the latest version for direct dependencies would solve most of these issues.

**Comment by ben-manes** [comment]
That is a recent change and has to do with Gradle’s non-resolvable parent configurations, if I understand the [commit](https://github.com/ben-manes/gradle-versions-plugin/commit/3274580adfe4ca03ea34c7442508a0f565f17d62) message.

We’re you thinking of transitive dependencies? Those are disabled for faster resolution so that only direct dependencies are resolved (meaning  transitive upgrades are not accounted for).

**Comment by bh-tt** [comment]
Yes, I was thinking of transitive dependencies, and I understand why you disabled those as those are much harder and error-prone to update.

**Comment by MartinX3** [comment]
I think he means that also older dependeny xml information are downloaded.
I use spring boot 3 and it downloads every 2.x version into the cache.
And of course not in parallel.

But the .+ versioning doesn't change this problem.

**Comment by ben-manes** [comment]
@MartinX3 I’ve seen Gradle do that because of resolution rules cause it to reject candidates, so it slowly walks the version history. That would be a build configuration mistake.

A similar case is a snapshot repository like the original author found. That’s easily solved using a repository content filter.

**Comment by MartinX3** [comment]
@ben-manes so is there something I need to change in all projects?
I never saw a different way of resolution.

**Comment by ben-manes** [comment]
You’d need to look at the info log when it does the resolution, seeing if a rejection reason was given. That’s either a resolution rule, our dsl helper for that (rejectVersionif), or maybe a dependency constraint.

When I’ve experienced it, I had a bug in my rule that rejected all versions so it failed to resolve a version. It was fairly obvious, a flipped condition or exception.

**Comment by ben-manes** [comment]
@MartinX3 also review this earlier [comment](https://github.com/ben-manes/gradle-versions-plugin/issues/505#issuecomment-1043664484) wrt the ivy revision flag.
